64 - New git-check action

mrsarm opened the pull request:
# Description

Ticket: [medic-conf#64](https://github.com/medic/medic-conf/issues/64)

### Implementation details

- The new action is called `check-git`, I tried to follow the convention name used in other action names, but I'm not 100% convinced of the name used, so I open to rename it.
- The implementation check local changes not commited and upstream changes, asking the user in both cases whether to continue or not.
- Add some refactors in the `exec-promise` module to be able to get the stout stream content that a command outputs, and also add the ability to silent or not the output regarless of the global log configuration. The PR adds also some unit tests to cover the `exec-promise` module.
- Pin "pluralize" dependency used in new module `git-exec` that was a dependency of dependency already.

Here is an example of the changes in action:

![medic-conf-issue-64-check-git](https://user-images.githubusercontent.com/1608415/87452922-302bb500-c5d8-11ea-93d7-3c627e066f5e.png)

# Code review items

- Readable: Concise, well named, follows the [style guide](https://github.com/medic/medic-docs/blob/master/development/style-guide.md), documented if necessary.
- Documented: Configuration and user documentation on [medic-docs](https://github.com/medic/medic-docs/)
- Tested: Unit and/or e2e where appropriate
- Internationalized: All user facing text
- Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in Changes.md.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

mrjones-plip commented:
As the least qualified of the requested reviewers, do y'all think it's worth it to have me be #5?  Also...so very many reviewers - is this a high risk ticket that you think needs lots of eyes on it?  Happy to help though if y'all think I'll be helpful!

mrsarm commented:
No this PR does not have nothing in special, I just added you and Jennifer because we are the new people in the town 😀 , so following the last Gareth suggestion (if I don't recall wrong) around PRs to add ourselves and also more experienced members from the team.

mrsarm commented:
Following up the comments above, I removed all the reviewers except one because now I understand I was not recalling well the reviewing process (and sorry for all my previous PRs where I also added a bunch of reviewers). I'll start from now only adding one reviewer unless need more oversight.

garethbowen commented on src/lib/exec-promise.js:
```diff
@@ -4,23 +4,23 @@ const log = require('../lib/log');
 // attacks by devious tech leads.
 
-module.exports = (...args) => new Promise((resolve, reject) => {
+module.exports = (log_level, ...args) => new Promise((resolve, reject) => {
```
Oof. I really want `log_level` to default so we don't have to specify it everywhere, but that doesn't play nicely with rest params. How would you feel about changing `...args` to an array so we can have a default log_level? eg:

```suggestion
module.exports = (args, logLevel=log.level) => new Promise((resolve, reject) => {
```

Then all the callers would pass an array of args rather than a variable number of parameters?

Also, use camel case for variables, eg: `logLevel` not `log_level`.

mrsarm replied:
Yes, it is a trade-off, because if we do that we also lost the ability to pass the args as separated parameters which is desired in this kind of functions. I'll do the change.

mrsarm replied:
Fixed

garethbowen commented on src/lib/exec-promise.js:
```diff
@@ -4,23 +4,23 @@ const log = require('../lib/log');
-    const stdio = log.level >= log.LEVEL_WARN  ? [ 'ignore',  'pipe',   'pipe'  ] :
-                  log.level >= log.LEVEL_ERROR ? [ 'ignore', 'ignore',  'pipe'  ] :
+    const stdio = log_level >= log.LEVEL_WARN  ? [ 'ignore',  'pipe',   'pipe'  ] :
+                  log_level >= log.LEVEL_ERROR ? [ 'ignore', 'ignore',  'pipe'  ] :
```
You didn't write this but while you're in here changing things, let's fix the nested ternary, eg: use `if/else`.

mrsarm replied:
Fixed

garethbowen commented:
Switching Stefan out for myself as I've already started reviewing  :)

garethbowen commented on test/lib/exec-promise.spec.js:
```diff
@@ -0,0 +1,29 @@
+      expect(err).to.match(/cmd-dont-exist\.sh: not found/);
+    }
+  });
+});
```
Explore mocking out `child_process` so we don't actually exec these commands.

Firstly the ideal unit test checks only the unit you're testing, rather than the whole system. This allows them to be as fast and stable as possible.

Secondly I'm not sure I want tests executing arbitrary code on my machine...

mrsarm replied:
Mocked

garethbowen commented on test/lib/git-exec.spec.js:
```diff
@@ -0,0 +1,9 @@
+  it('execute `git status` do not raise any error', async () => {
+    await git.status();
+  });
+});
```
As above, it'd be better to mock out the `exec` call so we can test the `git` unit without actually executing a child process on the machine. This will also mean you can test the `fetch` and `checkUpstream` functions in the `git` module which would otherwise involve a network request.

mrsarm replied:
Done.

garethbowen commented on src/lib/git-exec.js:
```diff
@@ -0,0 +1,42 @@
+ */
+module.exports.fetch = () => {
+  return exec(log.LEVEL_ERROR, GIT, 'fetch');
+};
```
I think this should be merged into the `checkUpstream` function because if you forget to call it then the `checkUpstream` won't work as expected. This abstracts the implementation away from the calling function.

mrsarm replied:
Merged.

mrsarm commented:
Hey @garethbowen , is it possible somebody using the tool without having a git repo? or not having yet an upstream repo configured? because in both cases this new action that is executed by default will make the action to fail, eg. if there is no git repo:

```
$ medic-conf --local
INFO Using local url from COUCH_URL environment variable: http://admin:****@localhost:5984/medic 
INFO Processing config in default.
INFO Actions:
     - check-git
     - compile-app-settings
     ...
     - upload-custom-translations
INFO Starting action: check-git…
fatal: not a git repository (or any of the parent directories): .git
ERROR fatal: not a git repository (or any of the parent directories): .git

$ _
```

do you want me to take into consideration these situations in the action code?

garethbowen commented:
As discussed in the standup: Great catch! We should definitely handle both of these situations gracefully with a warning rather than a fatal error. Thanks!

mrsarm commented:
@garethbowen I added one commit with the extra checks we discussed last week (whether the user has `git` installed and whether it has tracked the configuration that is going to push). Could you review it before move it again to test?

jonathanbataire commented:
Cool stuff here :100: 
its worked well for the most part except for this
so I've tried working with this scenario:
- Created a new branch
- make some changes
- add, commit and push
but i still get this warning
Also is `Are you sure?` supposed to be `Are you sure you want to continue?` like in the description screenshot :)
![Screenshot from 2020-08-12 11-51-39](https://user-images.githubusercontent.com/32832540/89995788-a7656f00-dc92-11ea-882a-02ea484450e8.png)

garethbowen commented on src/fn/check-git.js:
```diff
@@ -0,0 +1,35 @@
+module.exports = {
+  requiresInstance: false,
+  execute: async () => {
+    if (!environment.isProduction() && environment.instanceUrl) {
```
So if no `instanceUrl` is set we'll run the check? I'm wondering if we should skip it in this case, eg:

```suggestion
    if (!environment.isProduction() || !environment.instanceUrl) {
```

Can you think of an example of when we would want to check git but no instanceUrl is given?

mrsarm replied:
Maybe is unlikely but if the user call explicitly the actions to execute, eg.:

    $ mmconf check-git compress-pngs fetch-csvs-from-google-drive ...

And `check-git` is in that list explicity but no URL was provided (there are tasks that don't need to connect with the instance), that would be the case.

garethbowen replied:
Ok. We can always change it later if it's too annoying, but yes if they explicitly request it it should run!

garethbowen commented on src/lib/environment.js:
```diff
@@ -25,5 +30,16 @@ module.exports = {
+  isProduction() {
+    if (!this.instanceUrl) return false;
+    const hostname = url.parse(this.instanceUrl).hostname;
+    if (LOCAL_MATCHER.test(hostname)) return false;
```
Nitpick: for consistency use braces and new lines (this line, and line 40), eg:

```suggestion
    if (LOCAL_MATCHER.test(hostname)) {
      return false;
    }
```

I know this file has examples of the one line style, but it's inconsistent with the rest of the codebase.

Another option would be something like...

```suggestion
    return !LOCAL_MATCHER.test(hostname) &&
      !hostname.split('.').some(subdomain => DEV_MATCHER.test(subdomain));
```

garethbowen commented on src/lib/environment.js:
```diff
@@ -25,5 +30,16 @@ module.exports = {
+    if (!this.instanceUrl) return false;
+    const hostname = url.parse(this.instanceUrl).hostname;
+    if (LOCAL_MATCHER.test(hostname)) return false;
+    return !hostname.split('.').some(subdomain => DEV_MATCHER.test(subdomain));
```
I think this is a little unreliable, for example, if the domain name just happens to start with `dev-`, however it's good enough for this issue.

Please raise a new issue to improve this function either by requiring users to set it explicitly, or querying to the server, or something.

mrsarm replied:
Yes those kind of addresses are taken into account, eg. executing the expression against a URL like you mention returns `false` as expected:

```
$ node
> const DEV_MATCHER = /^(dev|test|staging)(-\w+|\d+)?$/;
undefined
> let hostname = 'test-v2.other.mm.org';
undefined
> !hostname.split('.').some(subdomain => DEV_MATCHER.test(subdomain));
false
```

mrsarm replied:
Actually the last test was wrong, the intention in the title targeted cases like these but it wasn't the right URL to test, checkout the fix at https://github.com/medic/medic-conf/pull/342/commits/5db0a68d36e96ee4f832d313a43c860d20189283

garethbowen commented on test/lib/environment.spec.js:
```diff
@@ -0,0 +1,66 @@
+      expect(environment.isProduction()).to.be.false;
+    });
+  });
+});
```
Great work!

garethbowen reviewed (changes requested):
Great job!

I've left a few minor or optional comments inline, but it's definitely almost there now!

garethbowen reviewed (approved):
Awesome work! I think this is ready for AT again - what do you think?

garethbowen commented:
@mrsarm This has passed AT. Can you please resolve the conflicts and merge?

mrsarm commented:
@garethbowen I need to add one more change discussed [here](https://github.com/medic/medic-conf/issues/64#issuecomment-697973860) in the ticket, I'm not working now on it but I expect to do soon.

garethbowen commented:
Ok. I've moved the issue back to "in progress". Thanks for the update!
